munihac/munihac.github.io
# Serve bootstrap from static webserver

# Goal of the Pull Request 

This Pull Request addresses a long standing issue of the Munihac website: loading resources from sources other than `munihac.de`. Apart from the iFrames, there were two offenders:
* https://cdn.jsdelivr.net/npm/bootstrap-icons (used to load bootstrap icons)
* https://ajax.googleapis.com/ajax/libs/jquery/ (used to load jquery)

Instead of serving `bootstrap-icons` and `jquery` from `munihac.de`, `index.html` would instruct the browser to fetch them from the aforementioned locations. With this Pull Request, everything is loaded from `munihac.de`. This change is important because:
1. User IP addresses are personal data, and
2. We cannot guarantee that IP addresses are handled properly by both https://cdn.jsdelivr.net and https://ajax.googleapis.com. By routing all the traffic to our own server, we ensure that we are not leaking personal data.

# Implementation

There are two main approaches to serve these dependencies from our own server:
1. Vendor-in the dependency (i.e. copy-paste the source code into our repository)
2. Consume the dependency as an `npm` package

I opted for the second option, because I consider that vendored dependencies are very hard to manage (if not simply impossible). Notably, `bootstrap` was vendored-in in the `js/` folder before this Pull Request -- I took the chance to also consume bootstrap as an npm package now.

To achieve my goal I did these fundamental changes:
1. I renamed `css/` to `assets/`, because I want to use the nodejs project to serve all sort of assets (styles, javascript, fonts) and not only css stylesheets
2. I updated the build pipeline of `assets/` (previously `css/`) to put all the static assets that we want to serve from `munihac.de` in a `dist/` folder. The build is now reproducible and contains all the necessary assets. This design also paves the way to a true build pipeline: ideally, the `dist/` folder would not be checked-in to the repository, but built in the CD pipeline.
3. I modified `site.hs` to compile the site correctly (taking into account the re-organization of the folders) and updated `default.html` to consume the static assets from their new location. The update to `site.hs` requires updating the `.cabal` file to include `filepath`.

# Other changes

* I upgraded from `bootstrap@5.1.x` to `bootstrap@5.3.3`. As part of the change, I had to import "maps" in `styles.css`.
* I uninstalled `jquery` and `@popperjs/core`. It is unclear to my why they were part of the project beforehand, but I don't consider them necessary -- they are included in the file `bootstrap.bundle.min.js` that we are serving.
* I deleted the vendored dependency `js/bootstrap.js`

# How to test the changes

* If you load the `munihac.de` version from the branch `munihac-2024`, you will see in the network tab that you are sending requests to a few different origins
* If you load this version, you will see in the network tab that you only send requests to `munihac.de` (with the exception of the google maps iFrame)
